# DATA ENGINEER
**Author**: Clarence Manzo (HCK-036)

This notebook contains codes to load the raw data and preprocess to ready-to-use dataset for data analysis and time series forecasting. The dataset contains transactional-level of Indonesian e-commerce order data from December 2023 – November 2025, including order-level information on product quantities, shipping fees, discounts, payment methods, and delivery destinations. Each row represents one completed or cancelled order.

# i. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import glob
import os

# ii. Data Loading

In [2]:
# Datasets path
folder_path = r".\dataset\RAW_PUBLIC_Dataset\RAW_PUBLIC"
file_paths = glob.glob(os.path.join(folder_path, "*.xlsx"))

# Initialize an array to store dataframes per month
df_list = []

# Read all files into a dataframe per monnth
for file in file_paths:
    df = pd.read_excel(file)
    df["source_file"] = os.path.basename(file)
    df_list.append(df)

# Combine all dataframes into single dataframe
df = pd.concat(df_list, ignore_index=True)

# View first 5 records
df.head(5)

,order_id,product_category,Status Pesanan,Alasan Pembatalan,Opsi Pengiriman,Waktu Pesanan Dibuat,Metode Pembayaran,Jumlah,Returned quantity,Total Diskon,...,Potongan Koin Shopee,Diskon Kartu Kredit,Ongkos Kirim Pengembalian Barang,Catatan dari Pembeli,Catatan,Username (Pembeli),Nama Penerima,No. Telepon,Alamat Pengiriman,Waktu Pesanan Selesai
0,ORD_0000001,Celengan,Selesai,NaN,Reguler (Cashless)-SPX Standard,2024-04-01 00:15,Saldo ShopeePay,1,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,ORD_0000001,Celengan,Selesai,NaN,Reguler (Cashless)-SPX Standard,2024-04-01 00:15,Saldo ShopeePay,1,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,ORD_0000002,Celengan,Selesai,NaN,Hemat Kargo-SPX Hemat,2024-04-01 01:47,COD (Bayar di Tempat),1,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,ORD_0000003,Celengan,Selesai,NaN,Hemat Kargo-SPX Hemat,2024-04-01 04:25,SeaBank Bayar Instan,1,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,ORD_0000004,Mangkok Sambal / Saus,Selesai,NaN,Hemat Kargo-SPX Hemat,2024-04-01 04:41,COD (Bayar di Tempat),2,0.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [3]:
# Print dataframe's shape (rows and columns)
df.shape

(26258, 50)

There's a total of **26258 records** of orders with **50 columns/attributes**.

In [4]:
# Show the missing values and data types from each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26258 entries, 0 to 26257
Data columns (total 50 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   order_id                                                      26258 non-null  object 
 1   product_category                                              26258 non-null  object 
 2   Status Pesanan                                                26258 non-null  object 
 3   Alasan Pembatalan                                             3525 non-null   object 
 4   Opsi Pengiriman                                               26258 non-null  object 
 5   Waktu Pesanan Dibuat                                          23824 non-null  object 
 6   Metode Pembayaran                                             26258 non-null  object 
 7   Jumlah                                                        26258

- Alasan Pembatalan: 22,733 missing values
- Waktu Pesanan Dibuat: 2,434 missing values
- Returned quantity: 1,335 missing values
- Waktu Pengiriman Diatur: 22,819 missing values
- Status Pembatalan/ Pengembalian: 26,248 missing values
- No. Resi: 24,936 missing values
- Antar ke counter/ pick-up: 24,936 missing values
- Pesanan Harus Dikirimkan Sebelum: 24,868 missing values
- Waktu Pembayaran Dilakukan: 24,788 missing values
- SKU Induk: 26,258 missing values (completely empty)
- Nomor Referensi SKU: 26,258 missing values (completely empty)
- Nama Variasi: 25,531 missing values
- Harga Awal: 24,788 missing values
- Harga Setelah Diskon: 24,788 missing values
- Total Harga Produk: 24,788 missing values
- Diskon Dari Penjual: 24,788 missing values
- Diskon Dari Shopee: 24,788 missing values
- Berat Produk: 24,788 missing values
- Jumlah Produk di Pesan: 24,788 missing values
- Voucher Ditanggung Penjual: 24,788 missing values
- Cashback Koin: 24,788 missing values
- Voucher Ditanggung Shopee: 24,788 missing values
- Paket Diskon: 24,788 missing values
- Paket Diskon (Diskon dari Shopee): 24,788 missing values
- Paket Diskon (Diskon dari Penjual): 24,788 missing values
- Potongan Koin Shopee: 24,788 missing values
- Diskon Kartu Kredit: 24,788 missing values
- Ongkos Kirim Pengembalian Barang: 24,788 missing values
- Catatan dari Pembeli: 26,200 missing values
- Catatan: 26,258 missing values (completely empty)
- Username (Pembeli): 24,799 missing values
- Nama Penerima: 24,788 missing values
- No. Telepon: 24,788 missing values
- Alamat Pengiriman: 24,788 missing values
- Waktu Pesanan Selesai: 24,987 missing values

In [5]:
# Print column's name
df.columns

Index(['order_id', 'product_category', 'Status Pesanan', 'Alasan Pembatalan',
       'Opsi Pengiriman', 'Waktu Pesanan Dibuat', 'Metode Pembayaran',
       'Jumlah', 'Returned quantity', 'Total Diskon', 'Total Berat',
       'Ongkos Kirim Dibayar oleh Pembeli',
       'Estimasi Potongan Biaya Pengiriman', 'Total Pembayaran',
       'Perkiraan Ongkos Kirim', 'Kota/Kabupaten', 'Provinsi', 'source_file',
       'Waktu Pengiriman Diatur', 'Status Pembatalan/ Pengembalian',
       'No. Resi', 'Antar ke counter/ pick-up',
       'Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)',
       'Waktu Pembayaran Dilakukan', 'SKU Induk', 'Nomor Referensi SKU',
       'Nama Variasi', 'Harga Awal', 'Harga Setelah Diskon',
       'Total Harga Produk', 'Diskon Dari Penjual', 'Diskon Dari Shopee',
       'Berat Produk', 'Jumlah Produk di Pesan', 'Voucher Ditanggung Penjual',
       'Cashback Koin', 'Voucher Ditanggung Shopee', 'Paket Diskon',
       'Paket Diskon (Diskon dari Shopee)',
       

In [6]:
# Check duplicate rows
print(df.duplicated().sum())

2462


There are total of **2462 duplicated data**, these data can be removed/dropped.

# iii. Data Preprocessing for Analysis

## 1. Drop Duplicate Data

In [7]:
# Drop duplicate rows
df = df.drop_duplicates()

# Check duplicate rows
print(df.duplicated().sum())

0


## 2. Drop Useless Columns

Columns with missing values above 20,000 data will be removed as it provides useless information. Missing value handling such as imputation will make the dataset biased. However, `Alasan Pembatalan` will be kept for data analysis purpose. Since most orders in this dataset are completed, missing values above 20,000 rows is expected in this column.

In [8]:
# Drop columns having almost completely missing values (above 20,000 except for 'Alasan Pembatalan')
cols_to_drop = [
    "Waktu Pengiriman Diatur",
    "Status Pembatalan/ Pengembalian",
    "No. Resi",
    "Antar ke counter/ pick-up",
    "Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)",
    "Waktu Pembayaran Dilakukan",
    "SKU Induk",
    "Nomor Referensi SKU",
    "Nama Variasi",
    "Harga Awal",
    "Harga Setelah Diskon",
    "Total Harga Produk",
    "Diskon Dari Penjual",
    "Diskon Dari Shopee",
    "Berat Produk",
    "Jumlah Produk di Pesan",
    "Voucher Ditanggung Penjual",
    "Cashback Koin",
    "Voucher Ditanggung Shopee",
    "Paket Diskon",
    "Paket Diskon (Diskon dari Shopee)",
    "Paket Diskon (Diskon dari Penjual)",
    "Potongan Koin Shopee",
    "Diskon Kartu Kredit",
    "Ongkos Kirim Pengembalian Barang",
    "Catatan dari Pembeli",
    "Catatan",
    "Username (Pembeli)",
    "Nama Penerima",
    "No. Telepon",
    "Alamat Pengiriman",
    "Waktu Pesanan Selesai"
]

# Drop the selected columns
df = df.drop(columns=cols_to_drop)

# Show the remaining columns
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23796 entries, 0 to 26257
Data columns (total 18 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   order_id                            23796 non-null  object 
 1   product_category                    23796 non-null  object 
 2   Status Pesanan                      23796 non-null  object 
 3   Alasan Pembatalan                   3233 non-null   object 
 4   Opsi Pengiriman                     23796 non-null  object 
 5   Waktu Pesanan Dibuat                21583 non-null  object 
 6   Metode Pembayaran                   23796 non-null  object 
 7   Jumlah                              23796 non-null  int64  
 8   Returned quantity                   22627 non-null  float64
 9   Total Diskon                        23796 non-null  float64
 10  Total Berat                         23796 non-null  object 
 11  Ongkos Kirim Dibayar oleh Pembeli   23796 non-

In [9]:
# Re-check the dataframe's shape
df.shape

(23796, 18)

## 3. Data Transformation

### 3.1 Convert Datatype

In [10]:
# Convert to datetime datatype
df['Waktu Pesanan Dibuat'] = pd.to_datetime(df['Waktu Pesanan Dibuat'])

# Remove ' gr' and convert to integer
df["Total Berat"] = df["Total Berat"].str.replace(" gr", "", regex=False).str.strip().astype(int)

# Remove '.' separator and convert to integer
df["Total Pembayaran"] = df["Total Pembayaran"].str.replace(".", "", regex=False).str.strip().astype(int)

In [11]:
# Re-check the datatypes
df.dtypes

order_id                                      object
product_category                              object
Status Pesanan                                object
Alasan Pembatalan                             object
Opsi Pengiriman                               object
Waktu Pesanan Dibuat                  datetime64[ns]
Metode Pembayaran                             object
Jumlah                                         int64
Returned quantity                            float64
Total Diskon                                 float64
Total Berat                                    int32
Ongkos Kirim Dibayar oleh Pembeli            float64
Estimasi Potongan Biaya Pengiriman           float64
Total Pembayaran                               int32
Perkiraan Ongkos Kirim                       float64
Kota/Kabupaten                                object
Provinsi                                      object
source_file                                   object
dtype: object

### 3.2 Cardinality Check

In [12]:
# Show the summary of object data distribution
df.describe(include='object')

,order_id,product_category,Status Pesanan,Alasan Pembatalan,Opsi Pengiriman,Metode Pembayaran,Kota/Kabupaten,Provinsi,source_file
count,23796,23796,23796,3233,23796,23796,23796,23796,23796
unique,20848,38,11,18,45,12,424,34,24
top,ORD_0019519,Celengan,Selesai,Dibatalkan oleh Pembeli. Alasan: Lainnya/ beru...,Hemat Kargo-SPX Hemat,COD (Bayar di Tempat),KAB. TANGERANG,JAWA BARAT,SeptemberSales2024_public.xlsx
freq,11,5998,20288,698,13957,12357,1336,7440,1466


In [13]:
# Define category columns and initialize array for unique values
cat_cols = list(df.select_dtypes(include='object').columns)
listItem = []

# Get unique values in each column
for col in cat_cols:
  listItem.append([col, df[col].nunique(), df[col].unique()])

# Display in dataframe
pd.set_option('display.max_colwidth', None)
pd.DataFrame(columns=['Column Name', 'Total Unique Values', 'Unique Values'], data=listItem)

,Column Name,Total Unique Values,Unique Values
0,order_id,20848,"[ORD_0000001, ORD_0000002, ORD_0000003, ORD_0000004, ORD_0000005, ORD_0000006, ORD_0000007, ORD_0000008, ORD_0000009, ORD_0000010, ORD_0000011, ORD_0000012, ORD_0000013, ORD_0000014, ORD_0000015, ORD_0000016, ORD_0000017, ORD_0000018, ORD_0000019, ORD_0000020, ORD_0000021, ORD_0000022, ORD_0000023, ORD_0000024, ORD_0000025, ORD_0000026, ORD_0000027, ORD_0000028, ORD_0000029, ORD_0000030, ORD_0000031, ORD_0000032, ORD_0000033, ORD_0000034, ORD_0000035, ORD_0000036, ORD_0000037, ORD_0000038, ORD_0000039, ORD_0000040, ORD_0000041, ORD_0000042, ORD_0000043, ORD_0000044, ORD_0000045, ORD_0000046, ORD_0000047, ORD_0000048, ORD_0000049, ORD_0000050, ORD_0000051, ORD_0000052, ORD_0000053, ORD_0000054, ORD_0000055, ORD_0000056, ORD_0000057, ORD_0000058, ORD_0000059, ORD_0000060, ORD_0000061, ORD_0000062, ORD_0000063, ORD_0000064, ORD_0000065, ORD_0000066, ORD_0000067, ORD_0000068, ORD_0000069, ORD_0000070, ORD_0000071, ORD_0000072, ORD_0000073, ORD_0000074, ORD_0000075, ORD_0000076, ORD_0000077, ORD_0000078, ORD_0000079, ORD_0000080, ORD_0000081, ORD_0000082, ORD_0000083, ORD_0000084, ORD_0000085, ORD_0000086, ORD_0000087, ORD_0000088, ORD_0000089, ORD_0000090, ORD_0000091, ORD_0000092, ORD_0000093, ORD_0000094, ORD_0000095, ORD_0000096, ORD_0000097, ORD_0000098, ORD_0000099, ORD_0000100, ...]"
1,product_category,38,"[Celengan, Mangkok Sambal / Saus, Other, Tempat Nasi, Keranjang, Nampan / Tray, Aksesoris Mandi, Toples / Sealware, Peralatan Kamar Mandi, Rak / Rak Serbaguna, Perlengkapan Packing, Plastik / Wadah Plastik, Stempel / Alat Kantor, Talenan, Saringan / Strainer, Botol / Gelas / Mug, Peralatan Makan, Bangku / Kursi Kecil, Aksesoris ID Card, Tempat Sampah, Lunch Box / Rantang, Mangkok, Seal / Baut / Roof, Piring, Gantungan Baju / Hanger, Pisau / Alat Potong, Teko / Jug, Sapu / Pembersih Lantai, Baskom / Mangkok Besar, Cobek / Ulekan, Aksesoris Pintu, Spatula, Pengolah Bumbu / Sayur, Sikat / Pembersih, Pot Tanaman / Bunga, Gayung, Perkakas, Aksesoris Motor]"
2,Status Pesanan,11,"[Selesai, Batal, Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-06., Sedang Dikirim, Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-05., Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-07., Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-09., Telah Dikirim, Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-08., Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-10., Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga 2025-12-11.]"
3,Alasan Pembatalan,18,"[nan, Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada, Dibatalkan oleh Pembeli. Alasan: Lainnya/ berubah pikiran, Dibatalkan secara otomatis oleh sistem. Alasan: Pengiriman gagal, Dibatalkan secara otomatis oleh sistem. Alasan: Penjual tidak mengatur pengiriman tepat waktu, Dibatalkan oleh Pembeli. Alasan: Need to change delivery address, Dibatalkan secara otomatis oleh sistem. Alasan: Pesanan belum dibayar, Dibatalkan secara otomatis oleh sistem. Alasan: Penjual gagal mengirimkan pesanan tepat waktu, Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman, Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan, Dibatalkan oleh Penjual. Alasan: Produk habis, Dibatalkan oleh Pembeli. Alasan: Menemukan yang lebih murah, Dibatalkan oleh Pembeli. Alasan: Penjual tidak responsif terhadap pertanyaan Pembeli, Dibatalkan oleh Pembeli. Alasan: Proses pembayaran sulit, Dibatalkan oleh Pembeli. Alasan: Lainnya, Dibatalkan secara otomatis oleh sistem. Alasan: Paket hilang di perjalanan. Kompensasi yang memenuhi syarat telah dikreditkan ke Saldo Penjual-mu., Dibatalkan oleh Pembeli. Alasan: Perlu mengubah Voucher, Dibatalkan oleh Pembeli. Alasan: Tidak ingin membeli lagi, Dibatalkan secara otomatis oleh sistem

- **order_id** is removed since it yields highest cardinality and doesn't provide any useful information.
- **Kota/Kabupaten** is removed since it has high cardinality and the information already covered in 'Provinsi'.
- **source_file** is removed since it has high cardinality and the source files are differentiate from each months which already covered in 'Waktu Pesanan Dibuat'.

In [14]:
# Drop selected columns
cols_to_drop = [
    "order_id",
    "Kota/Kabupaten",
    "source_file"
]

# Drop the selected columns
df = df.drop(columns=cols_to_drop)

Group the product categories into 6 major categories:  
Kitchen & Dining, Food Storage & Packaging, Bathroom & Cleaning, Home Organization & Living, Tools & Accessories, and Other

In [15]:
# Create mapping dictionary
category_mapping = {
    # Kitchen & Dining
    "Mangkok Sambal / Saus": "Kitchen & Dining",
    "Nampan / Tray": "Kitchen & Dining",
    "Talenan": "Kitchen & Dining",
    "Saringan / Strainer": "Kitchen & Dining",
    "Botol / Gelas / Mug": "Kitchen & Dining",
    "Peralatan Makan": "Kitchen & Dining",
    "Mangkok": "Kitchen & Dining",
    "Piring": "Kitchen & Dining",
    "Pisau / Alat Potong": "Kitchen & Dining",
    "Teko / Jug": "Kitchen & Dining",
    "Baskom / Mangkok Besar": "Kitchen & Dining",
    "Cobek / Ulekan": "Kitchen & Dining",
    "Spatula": "Kitchen & Dining",
    "Pengolah Bumbu / Sayur": "Kitchen & Dining",

    # Food Storage & Packaging
    "Tempat Nasi": "Food Storage & Packaging",
    "Toples / Sealware": "Food Storage & Packaging",
    "Perlengkapan Packing": "Food Storage & Packaging",
    "Plastik / Wadah Plastik": "Food Storage & Packaging",
    "Lunch Box / Rantang": "Food Storage & Packaging",

    # Bathroom & Cleaning
    "Aksesoris Mandi": "Bathroom & Cleaning",
    "Peralatan Kamar Mandi": "Bathroom & Cleaning",
    "Tempat Sampah": "Bathroom & Cleaning",
    "Sapu / Pembersih Lantai": "Bathroom & Cleaning",
    "Sikat / Pembersih": "Bathroom & Cleaning",
    "Gayung": "Bathroom & Cleaning",

    # Home Organization & Living
    "Celengan": "Home Organization & Living",
    "Rak / Rak Serbaguna": "Home Organization & Living",
    "Keranjang": "Home Organization & Living",
    "Bangku / Kursi Kecil": "Home Organization & Living",
    "Gantungan Baju / Hanger": "Home Organization & Living",
    "Pot Tanaman / Bunga": "Home Organization & Living",

    # Tools & Accessories
    "Stempel / Alat Kantor": "Tools & Accessories",
    "Seal / Baut / Roof": "Tools & Accessories",
    "Aksesoris Pintu": "Tools & Accessories",
    "Perkakas": "Tools & Accessories",
    "Aksesoris ID Card": "Tools & Accessories",
    "Aksesoris Motor": "Tools & Accessories",

    # Other
    "Other": "Other"
}

# Create new grouped column
df["product_category_grouped"] = df["product_category"].map(category_mapping)

# Check result
print(df["product_category_grouped"].value_counts())

product_category_grouped
Kitchen & Dining              8720
Home Organization & Living    7717
Tools & Accessories           3974
Food Storage & Packaging      1885
Other                          926
Bathroom & Cleaning            574
Name: count, dtype: int64


The original `Status Pesanan` column contains inconsistent and highly granular values.  
Specifically, multiple entries follow this pattern:

> "Pesanan diterima, namun Pembeli masih dapat mengajukan pengembalian hingga YYYY-MM-DD"

Although the dates differ, these values represent the same business state: The order has been received by the buyer.

Therefore, these values need to be merged into one single value which is **Pesanan Diterima**.

In [16]:
# Standardize 'Status Pesanan'
df["Status Pesanan"] = df["Status Pesanan"].replace(to_replace=r"Pesanan diterima.*", value="Pesanan Diterima", regex=True)

# Check result
print(df["Status Pesanan"].value_counts())

Status Pesanan
Selesai             20288
Batal                3233
Pesanan Diterima      172
Sedang Dikirim         71
Telah Dikirim          32
Name: count, dtype: int64


The original `Alasan Pembatalan` column contains long, inconsistent text strings that combine two different pieces of information:

1. **Who cancelled the order** (Pembeli, Penjual, or Sistem Otomatis)
2. **The specific reason for cancellation**

Separating these into two structured columns improves clarity and enables more precise analysis.

Reduce the text inconsistency as the original data contains:
- Mixed Indonesian and English reasons  
  (e.g., *"Need to change delivery address"* vs *"Perlu mengubah alamat pengiriman"*)
- Long explanatory system messages  
- Slightly different phrasings representing the same meaning  

Standardization ensures logically identical reasons are grouped together.

In [17]:
# Create Jenis Pembatalan
df["Jenis Pembatalan"] = np.select(
    [
        df["Alasan Pembatalan"].str.contains("Dibatalkan oleh Pembeli", na=False),
        df["Alasan Pembatalan"].str.contains("Dibatalkan oleh Penjual", na=False),
        df["Alasan Pembatalan"].str.contains("Dibatalkan secara otomatis oleh sistem", na=False),
    ],
    [
        "Pembeli",
        "Penjual",
        "Sistem Otomatis"
    ],
    default=np.nan
)

# Extract raw cancellation reasons
df["Alasan Pembatalan Clean"] = (
    df["Alasan Pembatalan"]
    .str.replace(r"Dibatalkan oleh Pembeli\. Alasan: ", "", regex=True)
    .str.replace(r"Dibatalkan oleh Penjual\. Alasan: ", "", regex=True)
    .str.replace(r"Dibatalkan secara otomatis oleh sistem\. Alasan: ", "", regex=True)
)

# Map the cancellation reasons
reason_mapping = {
    # PEMBELI
    "Ubah Pesanan yang Ada": "Ubah Pesanan yang Ada",
    "Perlu mengubah pesanan": "Perlu mengubah pesanan",
    "Lainnya/ berubah pikiran": "Lainnya/ berubah pikiran",
    "Need to change delivery address": "Perlu mengubah alamat pengiriman",
    "Perlu mengubah alamat pengiriman": "Perlu mengubah alamat pengiriman",
    "Menemukan yang lebih murah": "Menemukan yang lebih murah",
    "Penjual tidak responsif terhadap pertanyaan Pembeli": "Penjual tidak responsif",
    "Proses pembayaran sulit": "Proses pembayaran sulit",
    "Perlu mengubah Voucher": "Perlu mengubah voucher",
    "Tidak ingin membeli lagi": "Tidak ingin membeli lagi",
    "Lainnya": "Lainnya",

    # PENJUAL
    "Produk habis": "Produk habis",

    # SISTEM OTOMATIS
    "Paket hilang di perjalanan. Kompensasi yang memenuhi syarat telah dikreditkan ke Saldo Penjual-mu.": "Paket hilang di perjalanan",
    "Penjual gagal mengirimkan pesanan tepat waktu": "Penjual gagal mengirimkan pesanan tepat waktu",
    "Pengiriman gagal": "Pengiriman gagal",
    "Penjual tidak mengatur pengiriman tepat waktu": "Penjual tidak mengatur pengiriman tepat waktu",
}

# Create new column with the mapped cancellation reason
df["Alasan Pembatalan Final"] = df["Alasan Pembatalan Clean"].map(reason_mapping)
df.loc[(df["Jenis Pembatalan"] == "Sistem Otomatis") & (df["Alasan Pembatalan Final"].isna()), "Alasan Pembatalan Final"] = "Lainnya"

# Verify
print(df[["Jenis Pembatalan", "Alasan Pembatalan Final"]].drop_duplicates())

     Jenis Pembatalan                        Alasan Pembatalan Final
0                 nan                                            NaN
5             Pembeli                          Ubah Pesanan yang Ada
18            Pembeli                       Lainnya/ berubah pikiran
32    Sistem Otomatis                               Pengiriman gagal
46    Sistem Otomatis  Penjual tidak mengatur pengiriman tepat waktu
52            Pembeli               Perlu mengubah alamat pengiriman
84    Sistem Otomatis                                        Lainnya
93    Sistem Otomatis  Penjual gagal mengirimkan pesanan tepat waktu
384           Pembeli                         Perlu mengubah pesanan
849           Penjual                                   Produk habis
1107          Pembeli                     Menemukan yang lebih murah
1292          Pembeli                        Penjual tidak responsif
1313          Pembeli                        Proses pembayaran sulit
1439          Pembeli             

The original `Opsi Pengiriman` column contains highly granular and inconsistent shipping descriptions. Each value often combines multiple pieces of information in a single string, such as:
- Shipping provider (e.g., SPX, JNE, J&T, GoSend, GrabExpress)
- Service level (e.g., Reguler, Same Day, Instant, Kargo)
- Legacy labels (e.g., "Versi Lama")
- Platform default labels without explicit courier names

The column was transformed into:  
**1. Ekspedisi**: SPX, JNE, J&T, GoSend, GrabExpress, and Platform Default.  
**2. Metode Pengiriman**: Reguler, Hemat, Kargo, Express, Same Day Delivery, Instant Delivery, and Pick Up Point.

This transformation improves analytical clarity, reduces noise, and ensures the dataset is structured appropriately for reporting, performance analysis, and modeling.

In [18]:
# Map the shipping methods to the shipping brands
mapping_shipping = {
    'Reguler (Cashless)-SPX Standard': ('SPX', 'Reguler'),
    'Hemat Kargo-SPX Hemat': ('SPX', 'Hemat'),
    'Hemat Kargo-J&T Economy': ('J&T', 'Hemat'),
    'Instant (Versi Lama)-SPX Instant (Versi Lama)': ('SPX', 'Instant Delivery'),
    'Agen SPX Express': ('SPX', 'Pick Up Point'),
    'Instant (Versi Lama)-GrabExpress Instant (Versi Lama)': ('GrabExpress', 'Instant Delivery'),
    'Reguler (Cashless)-J&T Express': ('J&T', 'Reguler'),
    'Same Day-SPX Sameday': ('SPX', 'Same Day Delivery'),
    'Reguler (Cashless)-JNE Reguler': ('JNE', 'Reguler'),
    'Hemat Kargo': ('Platform Default', 'Hemat'),
    'Kargo-JNE Trucking (JTR)': ('JNE', 'Kargo'),
    'Kargo-J&T Cargo': ('J&T', 'Kargo'),
    'Next Day-JNE YES': ('JNE', 'Express'),
    'Instant (Versi Lama)': ('Platform Default', 'Instant Delivery'),
    'Same Day-GoSend Same Day': ('GoSend', 'Same Day Delivery'),
    'SPX Express Point': ('SPX', 'Pick Up Point'),
    'Instant (Versi Lama)-GoSend Instant (Versi Lama)': ('GoSend', 'Instant Delivery'),
    'Reguler (Cashless)': ('Platform Default', 'Reguler'),
    'GrabExpress Instant (Versi Lama)': ('GrabExpress', 'Instant Delivery'),
    'SPX Hemat': ('SPX', 'Hemat'),
    'SPX Standard': ('SPX', 'Reguler'),
    'J&T Cargo': ('J&T', 'Kargo'),
    'J&T Express': ('J&T', 'Express'),
    'GrabExpress Sameday': ('GrabExpress', 'Same Day Delivery'),
    'JNE Reguler': ('JNE', 'Reguler'),
    'GoSend Same Day': ('GoSend', 'Same Day Delivery'),
    'SPX Sameday': ('SPX', 'Same Day Delivery'),
    'Kargo': ('Platform Default', 'Kargo'),
    'Same Day': ('Platform Default', 'Same Day Delivery'),
    'Next Day': ('Platform Default', 'Express'),
    'Hemat Kargo-JNE Trucking (JTR)': ('JNE', 'Kargo'),
    'Hemat Kargo-J&T Cargo': ('J&T', 'Kargo'),
    'GoSend Instant (Versi Lama)': ('GoSend', 'Instant Delivery'),
    'SPX Instant (Versi Lama)': ('SPX', 'Instant Delivery'),
    'J&T Economy': ('J&T', 'Hemat'),
    'JNE Trucking (JTR)': ('JNE', 'Kargo'),
    'JNE YES': ('JNE', 'Express'),
    'Instant-SPX Instant': ('SPX', 'Instant Delivery'),
    'GrabExpress Instant': ('GrabExpress', 'Instant Delivery'),
    'Instant': ('Platform Default', 'Instant Delivery'),
    'Instant Prioritas-SPX Instant Prioritas': ('SPX', 'Instant Delivery'),
    'SPX Instant': ('SPX', 'Instant Delivery'),
    'GoSend Instant Prioritas': ('GoSend', 'Instant Delivery'),
    'SPX Instant Prioritas': ('SPX', 'Instant Delivery'),
    'Gosend Instant': ('GoSend', 'Instant Delivery')
}

# Create new column with the mapped shipping methods
df[['Ekspedisi', 'Metode Pengiriman']] = df['Opsi Pengiriman'].map(mapping_shipping).apply(pd.Series)

# Verify
print(df[['Ekspedisi', 'Metode Pengiriman']].drop_duplicates())

             Ekspedisi  Metode Pengiriman
0                  SPX            Reguler
2                  SPX              Hemat
11                 J&T              Hemat
25                 SPX   Instant Delivery
26                 SPX      Pick Up Point
46         GrabExpress   Instant Delivery
47                 J&T            Reguler
65                 SPX  Same Day Delivery
83                 JNE            Reguler
84    Platform Default              Hemat
98                 JNE              Kargo
196                J&T              Kargo
217                JNE            Express
384   Platform Default   Instant Delivery
413             GoSend  Same Day Delivery
435             GoSend   Instant Delivery
557   Platform Default            Reguler
779                J&T            Express
1021       GrabExpress  Same Day Delivery
1714  Platform Default              Kargo
2411  Platform Default  Same Day Delivery
2492  Platform Default            Express


### 3.3 Remove old columns, rename new columns, and format values

In [19]:
# Drop old columns
df = df.drop(columns=["product_category", "Alasan Pembatalan", 'Alasan Pembatalan Clean', 'Opsi Pengiriman'])

# Rename columns
df = df.rename(columns={
    "product_category_grouped": "Kategori Produk",
    "Alasan Pembatalan Final": "Alasan Pembatalan",
    "Returned quantity": "Returned Quantity"
})

# Capitalize each province word
df['Provinsi'] = df['Provinsi'].str.title()

# Change unit into thousands
cols = [
    "Total Diskon",
    "Ongkos Kirim Dibayar oleh Pembeli",
    "Estimasi Potongan Biaya Pengiriman",
    "Perkiraan Ongkos Kirim"
]
df[cols] = df[cols] * 1000

# Change weight unit from gr to kg
df['Total Berat'] = df['Total Berat'] / 1000

## 4. Handle Missing Values

In [20]:
# Check missing values
df.isna().sum()

Status Pesanan                            0
Waktu Pesanan Dibuat                   2213
Metode Pembayaran                         0
Jumlah                                    0
Returned Quantity                      1169
Total Diskon                              0
Total Berat                               0
Ongkos Kirim Dibayar oleh Pembeli         0
Estimasi Potongan Biaya Pengiriman        0
Total Pembayaran                          0
Perkiraan Ongkos Kirim                    0
Provinsi                                  0
Kategori Produk                           0
Jenis Pembatalan                          0
Alasan Pembatalan                     20563
Ekspedisi                                 0
Metode Pengiriman                         0
dtype: int64

Since time-series forecasting doesn't allow any missing date, missing `Waktu Pesanan Dibuat` should be removed. 

In [21]:
# Filter only for data that doesn't contain missing 'Waktu Pesanan Dibuat' values
df = df[df["Waktu Pesanan Dibuat"].notnull()]

In [22]:
# Check missing values pattern of returned quantity
df.groupby("Status Pesanan")["Returned Quantity"].apply(lambda x: x.isna().sum())

Status Pesanan
Batal               214
Pesanan Diterima      0
Sedang Dikirim        0
Selesai             955
Telah Dikirim         0
Name: Returned Quantity, dtype: int64

For completed orders, return quantity logically cannot exist if there is no cancellations. Therefore, missing `Returned quantity` will be imputed with 0.  
For cancelled orders, return quantity should be the same with the number of ordered quantity.

In [23]:
# Mask for Batal with missing return
mask_batal = (df["Status Pesanan"] == "Batal") & df["Returned Quantity"].isna()

# Mask for Selesai with missing return
mask_selesai = (df["Status Pesanan"] == "Selesai") & df["Returned Quantity"].isna()

# Impute cancelled orders with coressponding Jumlah
df.loc[mask_batal, "Returned Quantity"] = df.loc[mask_batal, "Jumlah"]

# Impute completed orders with 0
df.loc[mask_selesai, "Returned Quantity"] = 0

# Convert to integer
df["Returned Quantity"] = df["Returned Quantity"].astype(int)

In [24]:
# Re-Check missing values
df.isna().sum()

Status Pesanan                            0
Waktu Pesanan Dibuat                      0
Metode Pembayaran                         0
Jumlah                                    0
Returned Quantity                         0
Total Diskon                              0
Total Berat                               0
Ongkos Kirim Dibayar oleh Pembeli         0
Estimasi Potongan Biaya Pengiriman        0
Total Pembayaran                          0
Perkiraan Ongkos Kirim                    0
Provinsi                                  0
Kategori Produk                           0
Jenis Pembatalan                          0
Alasan Pembatalan                     18634
Ekspedisi                                 0
Metode Pengiriman                         0
dtype: int64

## 5. Create Features

In [25]:
# Create new features for net quantity sold and is weekend boolean
df["Jumlah Terjual Bersih"] = df["Jumlah"] - df["Returned Quantity"]
df["Weekend"] = (df["Waktu Pesanan Dibuat"].dt.weekday >= 5).astype(int)

# Display final dataframe for analysis
df.head(5)

,Status Pesanan,Waktu Pesanan Dibuat,Metode Pembayaran,Jumlah,Returned Quantity,Total Diskon,Total Berat,Ongkos Kirim Dibayar oleh Pembeli,Estimasi Potongan Biaya Pengiriman,Total Pembayaran,Perkiraan Ongkos Kirim,Provinsi,Kategori Produk,Jenis Pembatalan,Alasan Pembatalan,Ekspedisi,Metode Pengiriman,Jumlah Terjual Bersih,Weekend
0,Selesai,2024-04-01 00:15:00,Saldo ShopeePay,1,0,0.0,1.0,0.0,10000.0,38300,10000.0,Banten,Home Organization & Living,nan,NaN,SPX,Reguler,1,0
2,Selesai,2024-04-01 01:47:00,COD (Bayar di Tempat),1,0,0.0,0.5,0.0,14500.0,18576,14500.0,Jawa Tengah,Home Organization & Living,nan,NaN,SPX,Hemat,1,0
3,Selesai,2024-04-01 04:25:00,SeaBank Bayar Instan,1,0,0.0,0.5,0.0,8000.0,7069,8000.0,Jawa Barat,Home Organization & Living,nan,NaN,SPX,Hemat,1,0
4,Selesai,2024-04-01 04:41:00,COD (Bayar di Tempat),2,0,0.0,0.4,0.0,20000.0,32200,20000.0,Jambi,Kitchen & Dining,nan,NaN,SPX,Hemat,2,0
5,Batal,2024-04-01 06:12:00,COD (Bayar di Tempat),1,0,0.0,1.2,0.0,0.0,0,8000.0,Banten,Other,Pembeli,Ubah Pesanan yang Ada,SPX,Hemat,1,0


# iv. Data Preprocessing for Time Series Forecasting

Forecasting should be based on completed transactions since cancelled orders distort demand signal. Set dates as index for time-series forecasting purpose.

In [26]:
# Copy dataframe where 'Status Pesanan' is completed
df_forecast = df[df["Status Pesanan"] == "Selesai"].copy()

# Sort and set Waktu Pesanan Dibuat as index
df_forecast = df_forecast.sort_values("Waktu Pesanan Dibuat")
df_forecast = df_forecast.set_index("Waktu Pesanan Dibuat")

# Check dataframe's shape
df_forecast.shape

(18359, 18)

In [27]:
# Initialize a list of for storing datafranes
df_forecast_categories = {}

# Create dataframes aggregated by daily frequency per product categories
for cat in df_forecast["Kategori Produk"].unique():
    df_cat = df_forecast[df_forecast["Kategori Produk"] == cat].resample("D").agg({
        "Jumlah": "sum",
        "Returned Quantity": "sum",
        "Total Pembayaran": "sum",
        "Total Diskon": "sum",
        "Ongkos Kirim Dibayar oleh Pembeli": "sum",
        "Jumlah Terjual Bersih": "sum",
        "Weekend": "max"
    }).fillna(0)
    
    df_forecast_categories[cat] = df_cat

In [28]:
# Display Kitchen & Dining dataframe
df_forecast_categories["Kitchen & Dining"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,2,0,192890,0.0,86000.0,2,0.0
2023-12-02,14,0,467102,0.0,179000.0,14,1.0
2023-12-03,4,0,62724,0.0,9000.0,4,1.0
2023-12-04,5,0,241658,0.0,82000.0,5,0.0
2023-12-05,10,0,291600,0.0,102000.0,10,0.0
...,...,...,...,...,...,...,...
2025-11-23,9,0,169885,0.0,15000.0,9,1.0
2025-11-24,27,0,425226,0.0,0.0,27,0.0
2025-11-25,7,0,111691,1000.0,0.0,7,0.0


In [29]:
# Display Food Storage & Packaging dataframe
df_forecast_categories["Food Storage & Packaging"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,1,0,96445,0.0,43000.0,1,0.0
2023-12-02,0,0,0,0.0,0.0,0,0.0
2023-12-03,1,0,72240,0.0,33500.0,1,1.0
2023-12-04,1,0,74840,0.0,16000.0,1,0.0
2023-12-05,0,0,0,0.0,0.0,0,0.0
...,...,...,...,...,...,...,...
2025-11-21,1,0,94368,0.0,0.0,1,0.0
2025-11-22,2,0,44799,0.0,10800.0,2,1.0
2025-11-23,0,0,0,0.0,0.0,0,0.0


In [30]:
# Display Bathroom & Cleaning dataframe
df_forecast_categories["Bathroom & Cleaning"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-04,3,0,110840,0.0,16000.0,3,0.0
2023-12-05,1,0,18680,0.0,0.0,1,0.0
2023-12-06,0,0,0,0.0,0.0,0,0.0
2023-12-07,0,0,0,0.0,0.0,0,0.0
2023-12-08,0,0,0,0.0,0.0,0,0.0
...,...,...,...,...,...,...,...
2025-11-20,0,0,0,0.0,0.0,0,0.0
2025-11-21,0,0,0,0.0,0.0,0,0.0
2025-11-22,0,0,0,0.0,0.0,0,0.0


In [31]:
# Display Home Organization & Living dataframe
df_forecast_categories["Home Organization & Living"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,8,0,218100,0.0,32000.0,8,0.0
2023-12-02,2,0,71228,0.0,30500.0,2,1.0
2023-12-03,7,0,369855,0.0,160500.0,7,1.0
2023-12-04,15,0,507987,0.0,121000.0,15,0.0
2023-12-05,13,0,437864,0.0,77500.0,13,0.0
...,...,...,...,...,...,...,...
2025-11-21,7,0,214199,0.0,14000.0,7,0.0
2025-11-22,5,0,121434,0.0,0.0,5,1.0
2025-11-23,6,0,143803,0.0,0.0,6,1.0


In [32]:
# Display Tools & Accessories dataframe
df_forecast_categories["Tools & Accessories"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,3,0,65724,0.0,15000.0,3,0.0
2023-12-02,0,0,0,0.0,0.0,0,0.0
2023-12-03,1,0,1314000,0.0,0.0,1,1.0
2023-12-04,0,0,0,0.0,0.0,0,0.0
2023-12-05,3,0,62500,0.0,8000.0,3,0.0
...,...,...,...,...,...,...,...
2025-11-21,11,0,832899,1000.0,0.0,11,0.0
2025-11-22,17,0,238719,9000.0,3500.0,17,1.0
2025-11-23,7,0,78317,0.0,0.0,7,1.0


In [33]:
# Display Other dataframe
df_forecast_categories["Other"]

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,2,0,129370,0.0,51000.0,2,0.0
2023-12-02,0,0,0,0.0,0.0,0,0.0
2023-12-03,1,0,46800,0.0,3000.0,1,1.0
2023-12-04,1,0,20000,0.0,9000.0,1,0.0
2023-12-05,3,0,50000,0.0,8000.0,3,0.0
...,...,...,...,...,...,...,...
2025-11-14,0,0,0,0.0,0.0,0,0.0
2025-11-15,2,0,32470,0.0,0.0,2,1.0
2025-11-16,2,1,77900,0.0,0.0,1,1.0


In [34]:
# Create dataframe aggregated by daily frequency
df_forecast = df_forecast.resample("D").agg({
    "Jumlah": "sum",
    "Returned Quantity": "sum",
    "Total Pembayaran": "sum",
    "Total Diskon": "sum",
    "Ongkos Kirim Dibayar oleh Pembeli": "sum",
    "Jumlah Terjual Bersih": "sum",
    "Weekend": "max"
}).fillna(0)

# Display the dataframe
df_forecast

,Jumlah,Returned Quantity,Total Pembayaran,Total Diskon,Ongkos Kirim Dibayar oleh Pembeli,Jumlah Terjual Bersih,Weekend
Waktu Pesanan Dibuat,,,,,,,
2023-12-01,16,0,702529,0.0,227000.0,16,0.0
2023-12-02,16,0,538330,0.0,209500.0,16,1.0
2023-12-03,14,0,1865619,0.0,206000.0,14,1.0
2023-12-04,25,0,955325,0.0,244000.0,25,0.0
2023-12-05,30,0,860644,0.0,195500.0,30,0.0
...,...,...,...,...,...,...,...
2025-11-23,22,0,392005,0.0,15000.0,22,1.0
2025-11-24,35,0,531480,0.0,0.0,35,0.0
2025-11-25,13,1,211765,1000.0,13980.0,12,0.0


# v. Data Saving

In [35]:
# Save cleaned dataframe to a new CSV file
df.to_csv("cleaned_data_analysis.csv", index=False, encoding="utf-8")

# Save ready-to-use forecasting dataframe to a new CSV file
df_forecast.to_csv("forecast_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Kitchen & Dining"].to_csv("forecast_kitchen_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Food Storage & Packaging"].to_csv("forecast_storage_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Bathroom & Cleaning"].to_csv("forecast_bathroom_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Home Organization & Living"].to_csv("forecast_home_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Tools & Accessories"].to_csv("forecast_tools_data.csv", index=True, encoding="utf-8")
df_forecast_categories["Other"].to_csv("forecast_other_data.csv", index=True, encoding="utf-8")